src/main/scala/io/iohk/ethereum/mpt/MerklePatriciaTrie.scala

```diff
@@ -118,32 +114,32 @@ class MerklePatriciaTrie[K, V] private (private val rootHash: Option[Array[Byte]
   /**
     * Implicits
     */
-  private[mpt] implicit val nodeDec = new RLPDecoder[Node] {
-    override def decode(rlp: RLPEncodeable): Node = rlp match {
+  private[mpt] implicit val nodeDec = new RLPDecoder[MptNode] {
+    override def decode(rlp: RLPEncodeable): MptNode = rlp match {
       case RLPList(items@_*) if items.size == MerklePatriciaTrie.ListSize =>
-        val parsedChildren: Seq[Option[Either[Array[Byte], Node]]] = items.init.map {
+        val parsedChildren: Seq[Option[Either[ByteString, MptNode]]] = items.init.map {
           case list: RLPList => Some(Right(this.decode(list)))
           case RLPValue(bytes) =>
             if (bytes.isEmpty) None
-            else Some(Left(bytes))
+            else Some(Left(ByteString(bytes)))
         }
-        val terminatorAsArray: Array[Byte] = items.last
-        BranchNode(children = parsedChildren, terminator = if (terminatorAsArray.isEmpty) None else Some(terminatorAsArray), hashFn)
+        val terminatorAsArray: ByteString = items.last
+        BranchNode(children = parsedChildren, terminator = if (terminatorAsArray.isEmpty) None else Some(terminatorAsArray))
       case RLPList(items@_*) if items.size == MerklePatriciaTrie.PairSize =>
         val (key, isLeaf) = HexPrefix.decode(items.head)
-        if (isLeaf) LeafNode(key, items.last, hashFn)
+        if (isLeaf) LeafNode(ByteString(key), items.last)
         else {
           val next = items.last match {
             case list: RLPList => Right(this.decode(list))
-            case RLPValue(bytes) => Left(bytes)
+            case RLPValue(bytes) => Left(ByteString(bytes))
           }
-          ExtensionNode(key, next, hashFn)
+          ExtensionNode(ByteString(key), next)
         }
       case _ => throw new RuntimeException("Invalid Node")
     }
   }
 
-  private lazy val EmptyTrieHash = hashFn(encodeRLP(Array.emptyByteArray))
+  private lazy val EmptyTrieHash = Node.hashFn(encodeRLP(Array.emptyByteArray))
```

Comment:
Could it be that this is duplicated with `calculateEmptyRootHash`? Maybe we could leave just `calculateEmptyRootHash`